### PR TITLE
[META] Contributing Guide Revamp

### PR DESCRIPTION
**Pull request by hramos**
## Summary

This is the first step in revamping the Contribution Guide at `/CONTRIBUTING.md`. This work will require changes to the React Native website repository, as well as the React Native GitHub wiki. For review purposes, changes to copy that exists outside of this repository will be included within a `.github/wiki` directory in this PR, but they are not intended to be part of the final PR that gets merged.

## Changelog

[General] [Changed] - Contribution Guide Revamp

## Test Plan

N/A

### CODE REVIEW

File: .github/wiki/FirstPullRequest.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+You will now have a fork of React Native on GitHub at <https://github.com/your_username/react-native>. Next, you will grab a copy of the source code for your local machine. Open a shell and type the following commands:
+
+```bash
+git clone https://github.com/facebook/react-native.git
+git remote add fork https://github.com/your_username/react-native.git
+```
+
+> If the above seems new to you, do not be scared. You can access a shell through the Terminal application on macOS and linux, or PowerShell on Windows.
+
+A new `react-native/` directory will be created with the contents of the core React Native repository. This directory is actually a clone of the React Native git repository. It is set up with two remotes: `origin` for the upstream @facebook/react-native repository, and `fork` for the fork of React Native on your own GitHub account.
```
Comment (rickhanlonii) [inline .github/wiki/FirstPullRequest.md]:
```suggestion
A new `react-native/` directory will be created with the contents of the core React Native repository. This directory is called a "clone" of the React Native git repository. It is set up with two remotes: `origin` for the upstream @facebook/react-native repository, and `fork` for the fork of React Native on your own GitHub account.
```

File: .github/wiki/FirstPullRequest.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+```bash
+git checkout --branch my_feature_branch --track origin/master
+```
+
+## Chapter II: Implementing your Changes
+
+### 1. Make changes to the code
+
+You can now make any changes deemed necessary using your code editor of choice. Many popular editors integrate with source control in some way. You can also use `git status` and `git diff` on the command line to keep track of what has changed.
```
Comment (rickhanlonii) [inline .github/wiki/FirstPullRequest.md]: What do you think about having 4 sections here:

1. Make changes to the code
2. Test your changes
3. Lint your code
4. **View your changes (NEW)**

Then, section 1 can describe how to make changes (maybe with some links to resources) and section 4 can outline viewing the diff

File: .github/wiki/FirstPullRequest.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+### 2. Test your changes
+
+Make sure your changes are correct and do not introduce any test failures. You can learn more in [Running and Writing Tests](https://github.com/facebook/react-native/wiki/Tests).
+
+### 3. Lint your code
+
+We understand it can take a while to ramp up and get a sense of the style followed for each of the languages in use in the core React Native repository. Developers should not need to worry about minor nits, so whenever possible, we use tools that automate the process of rewriting your code to follow conventions.
+
+For example, we use [Prettier](https://prettier.io/) to format our JavaScript code. This saves you time and energy as you can let Prettier fix up any formatting issues automatically through its editor integrations, or by manually running `npm run prettier`. We also use a linter to catch styling issues that may exist in your code. You can check the status of your code styling by simply running `npm run lint`.
```
Comment (rickhanlonii) [inline .github/wiki/FirstPullRequest.md]:
```suggestion
We use [Prettier](https://prettier.io/) to format our JavaScript code. This saves you time and energy as you can let Prettier fix up any formatting issues automatically through its editor integrations, or by manually running `npm run prettier`. We also use a linter to catch styling issues that may exist in your code. You can check the status of your code styling by simply running `npm run lint`.
```

File: .github/wiki/FirstPullRequest.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+To learn more about coding conventions, refer to the [Style Guide](#style-guide).
+
+## Chapter III: Proposing your Changes
+
+### 1. Commit your changes
+
+Make sure to add your changes to version control using `git`:
+
+```bash
+git add <filename>
```
Comment (rickhanlonii) [inline .github/wiki/FirstPullRequest.md]:
```suggestion
git add .
```

File: .github/wiki/FirstPullRequest.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+remote: Create a pull request for 'your_feature_branch' on GitHub by visiting:
+remote:      https://github.com/your_username/react-native/pull/new/your_feature_branch
+remote:
+```
+
+Visit the provided URL to proceed to the next step.
+
+### 3. Create your pull request
+
+You are almost there! The next step is to fill out the pull request. Use a descriptive title that is not too long. Then, make sure to fill out all of the fields provided by the default pull request template:
```
Comment (rickhanlonii) [inline .github/wiki/FirstPullRequest.md]:
```suggestion
The final step is to submit the pull request! Be sure to use a descriptive title that is not too long, and make sure to fill out all of the fields provided by the default pull request template:
```

- Updated wording

File: .github/wiki/FirstPullRequest.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+**Summary:** Use this field to provide your motivation for sending this pull request. What are you fixing?
```
Comment (rickhanlonii) [inline .github/wiki/FirstPullRequest.md]: Just noting that by putting the template inline it will need to be maintained when we change the template, may consider linking to a new PR

File: .github/wiki/FirstPullRequest.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+**Changelog:** Help release maintainers write release notes by providing a short description of what will be changed should the pull request get merged.
+
+**Test Plan:** Let reviewers know how you tested your changes. Did you consider any edge cases? Which steps did you follow to make sure your changes have the desired effect? See [What is a Test Plan?][test-plan] to learn more.
+
+[test-plan]: https://medium.com/@martinkonicek/what-is-a-test-plan-8bfc840ec171
+
+### 4. Review and address feedback
+
+Keep an eye on any comments and review feedback left on your pull request on GitHub. Maintainers will do their best to provide constructive, actionable feedback to help get your changes ready to be merged into the core React Native repository.
```
Comment (rickhanlonii) [inline .github/wiki/FirstPullRequest.md]:
```suggestion
Keep an eye on any comments and review feedback left on your pull request on GitHub. Maintainers will do their best to provide constructive, actionable feedback to help get your changes ready to be merged into the core React Native repository.

### 5. Merged!

After your change has been approved and merged by a maintainer, you've successfully added your first change to React Native!
```

Just a suggestion for a final section to close on, my wording here could be improved

File: .github/wiki/Tests.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+[env-setup]: http://facebook.github.io/react-native/docs/getting-started
+
+
+### JavaScript Tests
+
+The simplest way to run the JavaScript test suite is by using the following command at the root of your React Native checkout:
+
+```bash
+npm test
```
Comment (rickhanlonii) [inline .github/wiki/Tests.md]:
```suggestion
yarn test
```

Comment (rickhanlonii) [reply .github/wiki/Tests.md]: Actually, just noticed that yarn is missing from the setup instructions as well

File: .github/wiki/Tests.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+npm run flow
+npm run lint
+```
+
+### iOS Tests
+
+To run the iOS tests, invoke the following script from the root of your React Native checkout:
+
+```bash
+./scripts/objc-test-ios.sh test
```
Comment (rickhanlonii) [inline .github/wiki/Tests.md]: Should these have npm scripts?

File: .github/wiki/Tests.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+You can run Detox end-to-end tests locally by [installing the Detox CLI](https://github.com/wix/Detox/blob/master/docs/Introduction.GettingStarted.md#step-1-install-dependencies) on macOS, then running the following in the command line:
+
+```bash
+npm run build-ios-e2e
+npm run test-ios-e2e
+```
+
+If you work on a component or API that isn't convered by a Detox test, please consider adding one. Detox tests are stored under [`RNTester/e2e/__tests__`](https://github.com/facebook/react-native/tree/master/RNTester/e2e/__tests__).
+
+### Continuous Testing
```
Comment (rickhanlonii) [inline .github/wiki/Tests.md]: This section is great, should it be mentioned in the contributing doc?

File: .github/wiki/Tests.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+[config-appveyor]: https://github.com/facebook/react-native/blob/master/.appveyor/config.yml
+[config-circleci]: https://github.com/facebook/react-native/blob/master/.circleci/config.yml
+
+There's another set of tests that run within Facebook's internal test infrastructure. Some of these tests are integration tests defined by internal consumers of React Native (e.g. unit tests for a React Native surface in the Facebook app). These tests run on every commit to the copy of React Native hosted on Facebook's source control. They also run when a pull request is imported to Facebook's source control. As it happens, should one of these tests fail, you'll need someone at Facebook to take a look. Since pull requests can only be imported by Facebook employees, whoever imported the pull request should be able to facilitate any details.
```
Comment (rickhanlonii) [inline .github/wiki/Tests.md]:
```suggestion
There's another set of tests that run within Facebook's internal test infrastructure. Some of these tests are integration tests defined by internal consumers of React Native (e.g. unit tests for a React Native surface in the Facebook app). These tests run on every commit to the copy of React Native hosted on Facebook's source control. They also run when a pull request is imported to Facebook's source control.

If one of these tests fail, you'll need someone at Facebook to take a look. Since pull requests can only be imported by Facebook employees, whoever imported the pull request should be able to facilitate any details.
```

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,35 +1,111 @@
...
+
+[facebook/react-native]: https://github.com/facebook/react-native
+<!-- END: Development Process -->
+
+### Repositories
+The main repository, <https://github.com/facebook/react-native>, contains the core React Native framework, and it is here where we keep track of bug reports and manage pull requests.
+
+There's a few other repositories you might want to familiarize yourself with:
+
+- The source code for the **React Native website**, including the documentation, is located at <https://github.com/facebook/react-native-website>
```
Comment (rickhanlonii) [inline CONTRIBUTING.md]:
```suggestion
- **React Native website** contains the source code for the website, including the documentation, is located at <https://github.com/facebook/react-native-website>
```

- Consistency with the other 2 bullet points

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -1,35 +1,111 @@
...
+- **Releases** are coordinated through the <https://github.com/react-native-community/react-native-releases> repository. This includes important documents such as the Changelog.
+- **Discussions** about the future of React Native take place in the <https://github.com/react-native-community/discussions-and-proposals> repository.
+
+Browsing through these repositories should provide some insight into how the React Native open source project is managed.
+<!-- END: Repositories. -->
+
+<!-- START: GitHub Issues -->
+# GitHub Issues
+
+We use GitHub issues to track public bugs. Please ensure your description is clear and has sufficient instructions to be able to reproduce the issue.
```
Comment (rickhanlonii) [inline CONTRIBUTING.md]:
```suggestion
We use GitHub issues to track bugs. Please ensure your description is clear and has sufficient instructions to be able to reproduce the issue.
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -1,35 +1,111 @@
...
+Facebook has a [bounty program](https://www.facebook.com/whitehat/) for the safe disclosure of security bugs. In those cases, please go through the process outlined on that page and do not file a public issue.
+<!-- END: GitHub Issues -->
+
+# Helping with Documentation
+
+The React Native documentation is hosted as part of the React Native website repository at https://github.com/facebook/react-native-website. The website itself is located at https://facebook.github.io/react-native and it is built using [Docusaurus](https://docusaurus.io/). If there's anything you'd like to change in the docs, you can get started by clicking on the "Edit" button located on the upper right of most pages in the website.
+
+If you are adding new functionality or introducing a change in behavior, we will ask you to update the documentation to reflect your changes.
+
+## Contributing to the Blog
```
Comment (rickhanlonii) [inline CONTRIBUTING.md]: Are we looking to actively include external blog posts on the website?

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -49,7 +125,31 @@ However, there are still some styles that the linter cannot pick up, notably in
 * If a method header doesn't fit on one line each argument goes on a separate line.
 * 100 character line length
 
-## License
+## Tests
+
+Tests help us prevent regressions from being introduced to the codebase. The GitHub repository is continuously tested using Circle and Appveyor, the results of which are available through the Checks functionality on [commits](https://github.com/facebook/react-native/commits/master) and pull requests. You can learn more about running and writing tests in the [Tests wiki](http://github.com/facebook/react-native/wiki/Tests).
```
Comment (rickhanlonii) [inline CONTRIBUTING.md]: Should this link to the nw testing doc as well?

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -49,7 +125,31 @@ However, there are still some styles that the linter cannot pick up, notably in
...
+<!-- END: Community Contributions -->
+
+<!-- START: Where to Get Help -->
+# Where to Get Help
+
+As you work on React Native, it is natural that sooner or later you may require help. In addition to the resources listed in [SUPPORT][support], people interested in contributing may take advantage of the following:
+
+* **Twitter**. The React Native team at Facebook has its own account at [**@reactnative**][at-reactnative], and the React Native Community uses [**@reactnativecomm**][at-reactnativecomm]. If you are stuck, or need help contributing, please do not hesitate to reach out.
+
+* **Proposals Repository**. If you are considering working on a feature large in scope, consider [creating a proposal first][meta]. The community is highly willing to help you figure out the right approach, and we'd be happy to help.
```
Comment (rickhanlonii) [inline CONTRIBUTING.md]:
```suggestion
* **Proposals Repository**. If you are considering working on a feature large in scope, consider [creating a proposal first][meta]. The community can help you figure out the right approach, and we'd be happy to help.
```

File: CONTRIBUTING.md (hunk 15 above)
Comment (hramos) [reply CONTRIBUTING.md]: Which one is that?

File: .github/wiki/Tests.md (hunk 8 above)
Comment (hramos) [reply .github/wiki/Tests.md]: Yarn is not used in the Getting Started guide intentionally, as it was written around the same time Yarn was released and the additional steps to install it were deemed extraneous as npm is already bundled with node.

I think we can update the website to use yarn, at which point we can do the same here. After all, we use yarn in our CI over npm.

File: .github/wiki/Tests.md (hunk 9 above)
Comment (hramos) [reply .github/wiki/Tests.md]: They don't at this time, but we should set it up. There's a lot of cleanup that can be done here, for example, there's three files: objc-test-ios.sh, objc-test-tvos.sh, and objc-test.sh. The latter should not be invoked directly, as the first two just set up the necessary envvars.

I actually wrote most of this doc over three months ago but have deferred adding it as I kept encountering areas where the DX could be improved, so I am leaning towards documenting how things are today even if its painful in some areas. Getting more eyes on it may help us get an idea which of these we should focus on fixing first.

File: CONTRIBUTING.md (hunk 14 above)
Comment (hramos) [reply CONTRIBUTING.md]: @cpojer might know

**Comment by hramos** [comment]
I've updated the Quip I based these on with the latest suggestions. Let's use that to work on any changes, and then I'll come back here and add them to the PR.

File: .github/wiki/Tests.md (hunk 8 above)
Comment (rickhanlonii) [reply .github/wiki/Tests.md]: Yeah that makes sense - the only risk is that new devs may run into some strange dependency issues with the difference in how deps are installed between yarn and npm 👍

File: .github/wiki/Tests.md (hunk 9 above)
Comment (rickhanlonii) [reply .github/wiki/Tests.md]: Ah yeah totally agree

File: CONTRIBUTING.md (hunk 15 above)
Comment (rickhanlonii) [reply CONTRIBUTING.md]: Nvm, I see that's the wiki/Tests.md linked here 👌

File: .github/wiki/How-To-Contribute.md [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+
+### Handling Pull Requests
+
+A pull request should consist of three main sections:
+
+* A summary. This helps us understand the motivation behind the changes.
+* A changelog. This helps us write the release notes. It also serves as a brief summary of your changes.
+* A test plan. This might be the most important part of your pull request. It lets us know that you have verified the change properly, and helps us understand if you've considered any edge cases.
+
+Any one pull request may require a deeper understanding of some area of React Native that you may not be familiar with. Even if you don't feel like you are the right person to review a pull request, you may still help by adding labels or asking the author for more information.
```
Comment (mokimo) [inline .github/wiki/How-To-Contribute.md]: This reads a bit weirdly `Any one pull request` (maybe it's just too late, trying to proofread 😄 )

Comment (hramos) [reply .github/wiki/How-To-Contribute.md]: I think removing "one" changes the meaning of the sentence, but I'm open to suggestions on how to reword this.

These are now in the wiki, which should make it easier for contributors to edit in place.
